sdk/core/amqp-common/src/requestResponseLink.ts

```diff
@@ -195,7 +195,7 @@ export class RequestResponseLink implements ReqResLink {
         const actionAfterTimeout = () => {
           timeOver = true;
           this.receiver.removeListener(ReceiverEvents.message, messageCallback);
-          const address = this.receiver.address || "address";
+          const address = this.receiver.source && this.receiver.source.address || "address";
```

Comment:
Are there more instances of this?

Reply:
It's possible - I was surprised it tried to access an undefined field which makes me wonder if this isn't supposed to happen :|

Reply:
If `this.receiver.source` is undefined, is the "receiver" considered to be in a usable state for further requests?
Should we consider "closing" the `RequestResponseLink` in such a case?
Maybe log an issue in rhea repo to consult when does `source` get to an `undefined` state and what are the suggested next steps?

Reply:
The short:

Looking at this more I think what I have above is our best defensive maneuver to remediate this. This particular route will end up causing the promise to reject, and give the user a reason to retry which I think is their best option.

The long:

I've been through the rhea code and amqp-common. There are a few hops to get there but the data that we're returning as `this.receiver.source` is actually an alias to the 'link.remote.attach.source' field which is stored as part of receiving the `attach` frame from Service Bus - this data is just stored directly so I think it's a faithful and fairly dumb copy of what's in the frame itself.

After thinking about this the only way I can see this field being undefined is if it was sent to us that way - meaning we received the attach frame but the attach.source field is empty (possibly this is a bug in rhea - unsure). Both customer bugs that we've seen that have exhibited this behavior also seem to involve disconnection/connection events (in one case I believe we narrowed it down to the Service Bus being upgraded on the backend and in the other the customer is doing a manual failover). So we might have something we can attempt to stress test against.

With that said, these are incredibly hard to replicate so I think our best fix is to eliminate the obvious crash (using what I have above). 

rhea code related to this field:
* [aliasing](https://github.com/amqp/rhea/blob/5a94367cc0069de6f69f102f059a69e6ead55f45/lib/link.js#L231)
* [setting of the remote.attach field](https://github.com/amqp/rhea/blob/5a94367cc0069de6f69f102f059a69e6ead55f45/lib/link.js#L142)

> If this.receiver.source is undefined, is the "receiver" considered to be in a usable state for further requests?
Should we consider "closing" the RequestResponseLink in such a case?

It seems logical to me. My only hesitation at this point is that I don't want to introduce more code into the code path while we're in such an odd state. It might actually be _worse_ but without any way to reproduce this scenario I'd like to "play it safe" if possible.

One additional thing I think we could do is reject the promise in setTimeout() with a non-retryable error when we detect that .source is incorrect, so at least we stop unnecessary retries.